src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs

```diff
@@ -373,37 +478,85 @@ private TValue TryAddOrGetExisting(TValue value, out bool addedValue)
             // replaced by expansion. If it has, we need to restart and write to the new array.
-            if (_newHashTable != hashTableLocal)
+            if (Interlocked.CompareExchange(ref _newHashTable, hashTableLocal, hashTableLocal) != hashTableLocal)
             {
+                if (!TryWriteAbortNullToLocation(hashTableLocal, tableIndex))
+                {
+                    // Indicates that the overwrite of the sentinel didn't work. That shouldn't be possible, so failfast
+                    Environment.FailFast("Unable to overwrite sentinel with null");
+                }
                 // Pulse the lock so we don't spin during an expansion
                 lock(this) { }
                 Interlocked.Decrement(ref _reserve);
                 return null;
             }
 
+            if (!TryWriteValueToLocation(value, hashTableLocal, tableIndex))
+            {
+                // Indicates that the overwrite of the sentinel didn't work. That shouldn't be possible, so failfast
+                Environment.FailFast("Unable to overwrite sentinel with value");
+            }
+
             // If the write succeeded, increment _count
             Interlocked.Increment(ref _count);
             return value;
         }
 
         /// <summary>
-        /// Attampts to write a value into the table. May fail if another value has been added.
+        /// Attempts to write a the sentinel into the table. May fail if another value has been added.
+        /// </summary>
+        /// <returns>True if the sentinel was successfully written</returns>
+        private bool TryWriteSentinelToLocation(TValue[] hashTableLocal, int tableIndex)
+        {
+            // Add to hash, use a volatile write to ensure that
+            // the contents of the value are fully published to all
+            // threads before adding to the hashtable
+            if (Interlocked.CompareExchange(ref hashTableLocal[tableIndex], _entryInProcessOfWritingSentinel, null) == null)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Attempts to write a value into the table. Should never fail as the sentinel should be the only
+        /// entry that can be in the table at this point
         /// </summary>
         /// <returns>True if the value was successfully written</returns>
         private bool TryWriteValueToLocation(TValue value, TValue[] hashTableLocal, int tableIndex)
         {
             // Add to hash, use a volatile write to ensure that
             // the contents of the value are fully published to all
             // threads before adding to the hashtable
-            if (Interlocked.CompareExchange(ref hashTableLocal[tableIndex], value, null) == null)
+            if (Interlocked.CompareExchange(ref hashTableLocal[tableIndex], value, _entryInProcessOfWritingSentinel) == _entryInProcessOfWritingSentinel)
```

Comment:
>Interlocked.CompareExchange [](start = 16, length = 27)

Same here as in `TryWriteAbortNullToLocation`. Could be an ordinary write. If want to assert for the original value, could just read it before writing. You've "claimed" the slot by writing sentinel to it so should be the only writer of the actual value or null.

And if the exclusivity is violated, ordinary read/write will catch that too.